### PR TITLE
12671 querying by user attribute no longer forces cases insensitivity

### LINKED ISSUE #12671: Slow user query by attribute
### Describe the bug

If querying users by attribute values the response is really slow. 

On further investigation the query is missing the database index because attribute name and value is queried by lower value, eg. lower(name) and lower(value).

Either remove lower and match the actual value (which seems to be the more reasonable approach) or add a matching new index to the attribute table: 
``` CREATE INDEX index ON user_attribute(lower(name), lower(value)); ```

### Version

18.0.1

### Expected behavior

Query by attribute returns in a reasonable time.

### Actual behavior

Query takes really long and produces a lot of load on the database.

### How to Reproduce?

Just query by attribute.

### Anything else?

Adding a max value to the query totally breaks the request. (Timeouts)

### PR DESCRIPTION
**Pull request by sschu**
Fixes #12671

### CODE REVIEW

**Comment by sschu** [comment]
This definitely also needs a mention in the change notes. Furthermore, a test might be good but I haven't found an existing test and I am also not sure how to test this since the behaviour is now collation-dependant and I don't know if the collation used in the tests is case-sensitive or not.
@hmlnarik @ahus1 WDYT?

**Comment by ahus1** [comment]
@sschu - thank you for providing this PR. 

There is `UserModelTest#testSearchByUserAttribute()` which could be extended to show that using a different case doesn't show up the users any more. 

An entry to the release notes and the migration guide is definitely needed. 

As the old behavior has been around for quite some time, I suggest to go even one step further to add a configuration via a provider configuration to `JpaUserProvider` so customers to restore the old behavior while they are migrating their clients. 

While we need an automated test for the new behavior, IMHO we don't need an automated test for the old behavior when we test it manually.

**Comment by sschu** [comment]
@ahus1 Thanks for the pointers! Thinking about the provider configuration, there is a related issue I wanted to bring up and maybe it makes sense to already discuss it here: case-sensititivity/insensitivity of DB collations.

Currently, Keycloak forces case insensitivity in the JPAUserProvider for custom attributes, for firstname and lastname and also in the policy/resource store and some places in the new store. This generally means that in the query there is as "WHERE LOWER(COLUMN)...".

PostgreSQL can still use an index in case of this query if the index is specifically created on LOWER(COLUMN). MS SQL does not support this. In MS SQL one must add a computed column to a table and also modify the query to use that column. I have not checked how this is in other databases but the current mechanism makes querying on MS SQL really inefficient.

On the other hand, there are case-sensitive and case-insensitive DB collations. Keycloak forcing lowercase comparisons when the collation is already case-insensitive does not really add value but potentially kills index usage. To my knowledge, only PostgreSQL has a case-sensitive default collation. MySQL, MariaDB, CockroachDB, MS SQL are all case-insensitive by default. I could not get a definite answer for Oracle.

Therefore, it would be good if users could specify whether Keycloak should actually force case-insensitivity and maybe even make a default choice depending on the database used.

On a related note, having indexes on realm+firstname and realm+lastname would be helpful to speed up the general search and querying specifically for these parameters. 

Bonus points if these indexes would automagically be on "LOWER(FIRSTNAME)" in case of databases like PostgreSQL that can use an index on "LOWER(COLUMN)" automatically. But maybe its better not to overload this issue now. I can also open a separate discussion for that as I initially intended to to.

**Comment by ahus1** [comment]
> Therefore, it would be good if users could specify whether Keycloak should actually force case-insensitivity and maybe even make a default choice depending on the database used.

I hear you when you describe differences between database. At the same time "making it configurable for the user" is to me a slippery slope, as it is sounds like "we don't know how this should work for a user". Making this configurable would increase the need to test this, and would increase the complexity for users to install Keycloak. 

I'd rather want us to be brave enough to choose an opinionated user experience first (either case insensitive or not case insensitive), and then make it work across all supported databases as needed (even if that requires a computed column on some databases, and a separate query - we already have the means to do that). 

For attribute names, I don't see the need for case-insensitive search. I might see a case for attribute values, but looking at the parent issue there seems to be consensus to not have case sensitive search for attribute values. 

So TL;DR: Let's make it behave nicely from a user's point of view first, and then align the technology in the background to make it so!

**Comment by ahus1** [comment]
Talking about case insensitive search: is uses exact case: firstname and username is using case-insensitive search, and username and email are converted to lowercase before written to the database, so this is fine as well. 

So shouldn't we be more consistent with attribute as well - keys case-sensitive, values case-insensitive?

https://github.com/keycloak/keycloak/blob/29b67fc8dfcd389b048672c1a3a7cf421efbc3a8/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L900-L925

**Comment by sschu** [comment]
> > Therefore, it would be good if users could specify whether Keycloak should actually force case-insensitivity and maybe even make a default choice depending on the database used.
> 
> I hear you when you describe differences between database. At the same time "making it configurable for the user" is to me a slippery slope, as it is sounds like "we don't know how this should work for a user". Making this configurable would increase the need to test this, and would increase the complexity for users to install Keycloak.
> 
> I'd rather want us to be brave enough to choose an opinionated user experience first (either case insensitive or not case insensitive), and then make it work across all supported databases as needed (even if that requires a computed column on some databases, and a separate query - we already have the means to do that).
> 
> For attribute names, I don't see the need for case-insensitive search. I might see a case for attribute values, but looking at the parent issue there seems to be consensus to not have case sensitive search for attribute values.
> 
> So TL;DR: Let's make it behave nicely from a user's point of view first, and then align the technology in the background to make it so!

I get your points. Therefore, I would assume the intention of Keycloak is to have search case-insensitive - because that is what the code looks like.

The most efficient solution would be to have case-insensitivity in the DB collation. However, the collation can typically only be specified at DB creation time. I don't think it is realistic to create a solution that only works if Keycloak is set up from scratch. Furthermore, there could be reasons for people to pick their own collations like using non-Latin character sets.

So I think the best would probably be if Keycloak somehow detects if the collation in use is case-insensitive or not and depending on that creates matching indexes and emits matching queries? And if there is no reliable way to detect it, having a default per DB and a way to override is the poor man's version of it?

**Comment by sschu** [comment]
> Talking about case insensitive search: is uses exact case: firstname and username is using case-insensitive search, and username and email are converted to lowercase before written to the database, so this is fine as well.
> 
> So shouldn't we be more consistent with attribute as well - keys case-sensitive, values case-insensitive?
> 
> https://github.com/keycloak/keycloak/blob/29b67fc8dfcd389b048672c1a3a7cf421efbc3a8/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L900-L925

I'll adapt this PR accordingly.

**Comment by sschu** [comment]
So I was really scratching my ahead about this one till I finally realized that the code being tested here is actually a different one, it is: https://github.com/keycloak/keycloak/blob/a7e014fbb7f0c181f73000ccd63480c416000abd/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L759-L766 and the contained named query.

The code was actually inconsistent with the search functionality as id did not force any case-insensitivity. So I changed that to make comparison of user attribute values case-insensitive.

However, that also raises the question about how this should be in other UserStorageProviders like LDAP, MapStorage etc. Should Keycloak rely on these providers to provider to be either case-sensitive or not?  Maybe the providers should actually report that and then Keycloak can adapt its logic? Something in that direction?

WDYT @ahus1 ?

**Comment by jiehao2019** [comment]
We have encountered a similar problem where we are using a MySQL database and it cannot use indexes, resulting in slow queries. 
In my opinion, most of the people who use keycloak are technical personnel such as programmers and devops. Most of them are actually accustomed to configuring based on the needs of the project. As a user, I tend to agree with the viewpoint of @sschu . I don't know if there will be further discussion in the community about this issue.

**Comment by pierre-isb** [comment]
I agree with @Akireken : attribute values being indexed, it might be interesting to have the ability to perform exact queries on attribute values to get faster results on large databases.

**Comment by maxime-bouchee** [comment]
Hello !, I have the same issue on my project, it would be cool to have this feature :smile:

**Comment by PabloLec** [comment]
Hello, same problem on our project: Searches for uppercase values mean the index is not used, causing extremely slow queries. Moreover, we use SQL Server which does not allow creating indexes with functions like lower(). From my perspective, the solution doesn't matter, but if the goal is to be case insensitive, why not enforce insertion in lower()? This would require only one operation, at the insertion, and would make the search lighter and avoid any indexing issues. That said, obviously, this would have implications on existing data.

**Comment by sschu** [comment]
@PabloLec You could have a second computed column with the value in lower case, yes. But the original value needs to retain its casing since for example names typically start with upper case letters. Also, you have the existing data as you mentioned.

**Comment by sschu** [comment]
Correct, the query Keycloak is doing also needs to use the computed column.

**Comment by sschu** [comment]
> @sschu I'm also interested in moving forward with this one. Can you please summarize the main points blocking this PR and the issue from being fixed?
> 
> What I understand from the discussions is:
> 
> * `username` and `email` are case sensitive and they are not affected as they are persisted in lower-case.
Correct. No issue here.
> * other attributes (including `firsName` and `lastName`) are persisted without forcing a case and keeping their original value
Correct.
> * it does not make sense to use the `lower` function on the column name for these other attributes. Including the `value`, but your PR is not changing this.
Correct.
> * queries are using the `lower` function in the attribute name and that blocks the database from properly using the index as it needs to go over all rows, run `lower` on the name, and then run the comparison.
This is AFAIK true for the index that Keycloak generates which is directly on the attribute name as the first element. Depending on the database, you can still make it work. In Postgres, you can create an index on a computed function like `create index on lower(user_attributes.name)` and Postgres will automatically use it when a query contains `where lower(user_attributes.name)` is used. This way people can speed this up by just adding another index. In MS SQL, you would have to add a computed column to the table itself that and then you can add an index on that computed column. But then the query also needs to look different as it has to target the computed column.
> * most databases are case-insensitive and the `lower` function can be omitted from queries to leverage database indexes
This completely depends on the collations being used. The default collation for Postgres is case-sensitive so you need all the `lower()` stuff if you want case-insensitive searches. The default collation for MS SQL is case-insensitive so all this `lower()` is not necessary. It is even harmful as it prevents indexes from being used.
> 
> Why not just remove the lower function from queries? And only use case-insensitive queries when a provider option tells so, as you suggested?
The counter-argument by @ahus1 and @stianst to this was that it adds yet another option which is not easy to handle for people with less DB knowledge. Keycloak could be smarter about this. If Keycloak wants comparisons for certain fields to be case-insensitive, there are two ways:
a) Try to detect the collation and do different things: collation is case-insensitive? add an index directly on the column and don't use `lower()` in the query (e.g. for MS SQL) collation is case-sensitive? add an index on the computed column and use `lower()` (e.g. for Postgres). This is the most efficient option.
b) Add a separate column with lowercase values, put an index on that and use that column in the query. This will work on all databases but be less efficient..
> 
> For PostgreSQL, we could even create the suggested index with `lower` by default.
True, see option a)

What I tried to do in this PR is change Keycloak's opinion if comparisons on attribute names should always be case-insensitive. By removing the `lower(user_attributes.name)`, the existing index will be used on all databases and if the comparison is actually case-insensitive or not depends on the database.

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserEntity.java [hunk 1]
```diff
@@ -50,7 +50,7 @@
         @NamedQuery(name="getRealmUserCount", query="select count(u) from UserEntity u where u.realmId = :realmId"),
         @NamedQuery(name="getRealmUserCountExcludeServiceAccount", query="select count(u) from UserEntity u where u.realmId = :realmId and (u.serviceAccountClientLink is null)"),
         @NamedQuery(name="getRealmUsersByAttributeNameAndValue", query="select u from UserEntity u join u.attributes attr " +
-                "where u.realmId = :realmId and attr.name = :name and attr.value = :value"),
+                "where u.realmId = :realmId and attr.name = :name and lower(attr.value) = :value"),
```
Comment (pedroigor) [inline model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserEntity.java:53]: Aren't we changing the expectation around the default collation used by the database? As we are basically querying custom attributes, I might not want `foo` the same as `FoO`. And that is how we have been working.

**Review by pedroigor — DISMISSED** [review]

**Review by pedroigor — COMMENTED** [review]
Sorry, the wrong option was choosen. Still have a comment :)

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java [hunk 2]
```diff
@@ -992,7 +992,7 @@ private List<Predicate> predicates(Map<String, String> attributes, Root<UserEnti
                     Join<UserEntity, UserAttributeEntity> attributesJoin = root.join("attributes", JoinType.LEFT);
 
                     attributePredicates.add(builder.and(
-                            builder.equal(builder.lower(attributesJoin.get("name")), key.toLowerCase()),
+                            builder.equal(attributesJoin.get("name"), key),
                             builder.equal(builder.lower(attributesJoin.get("value")), value.toLowerCase())));
```
Comment (pedroigor) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java:996]: Perhaps you did lower in the other method to get consistent with this one?

**Comment by pedroigor** [comment]
@sschu Thanks. Based on your reply I would propose the following:

* Let us merge this PR and make sure the `name` is not using the `lower` function. Basically, for the reasons you mentioned at the bottom of your comment. We can also remove the `lower` function around the `value` to keep the two methods `searchForUserByUserAttributeStream` and `predicates` consistent and use case-sensitive searches.
* Create a follow-up issue to come up with a smarter way of deciding whether we should use case-sensitive and case-insensitive queries.

The idea is to have improvements in increments whereas the changes herein look like a great start to evaluate and see how it solves the original problem.

**Comment by sschu** [comment]
@pedroigor I agree with your proposal. I'll rewrite this PR to solve the merge conflicts.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java (hunk 2 above)
Comment (sschu) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java]: Exactly.

**Comment by sschu** [comment]
@pedroigor After quite some code digging I decided to strip this down to the bare minimum. I only made user attribute key comparisons not force case-insensitivity so the index will be used. I could not add a test for this change since my change does not mean user attribute key comparisons are case-sensitive, that still depends on the DB used.

I changed one existing test that searches for normal length user attributes just to make sure its value comparison is case-insensitive (this is the current behaviour and it is explicitly tested for the new extended length attributes but not for old shorter attributes).

I would defer the discussion around consistency of forcing case-insensitivity in the UserStorageProviders to a separate issue. There are quite some affected places (JpaUserProvider, LdapStorageProvider, JpaUserFederatedStorageProvider) and changes there might break something. So it's worth to have a dedicated discussion around this.

**Review by pedroigor — DISMISSED** [review]

**Review by pedroigor — COMMENTED** [review]
@sschu Thanks. 

Sure, let's start with something and have follow-ups on more improvements. The changes here work for me.

I just realized that we probably want to add some notes in the release notes and upgrading guide to mention that queries on attributes are no longer using the `lower` function on the `name` so people can adapt (if needed) any custom index they have created.

Does it make sense to you?

**Comment by sschu** [comment]
@pedroigor Makes sense, thanks. I added some notes to both documents.

**Review by sguilhen — APPROVED** [review]
Took me a while to go through the whole discussion history, but finally made it. I agree with the proposal to keep this one as simple as possible and handle different case-sensitive searches after a separate discussion.

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@sschu Thanks.
